discoverd2/server/dns.go

```diff
@@ -0,0 +1,388 @@
+package server
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/miekg/dns"
+	"github.com/flynn/flynn/discoverd2/client"
+	"github.com/flynn/flynn/pkg/random"
+)
+
+type DNSStore interface {
+	Get(string) []*discoverd.Instance
+	GetLeader(string) *discoverd.Instance
+}
+
+type DNSServer struct {
+	UDPAddr   string
+	TCPAddr   string
+	Store     DNSStore
+	Domain    string
+	Recursors []string
+
+	servers []*dns.Server
+}
+
+const maxUDPRecords = 3
+const dnsDomain = "discoverd."
+
+func (srv *DNSServer) ListenAndServe() error {
+	if srv.Store == nil {
+		panic("missing Store")
+	}
+	if srv.Domain == "" {
+		srv.Domain = dnsDomain
+	}
+	if err := srv.validateRecursors(); err != nil {
+		return err
+	}
+
+	api := dnsAPI{srv}
+	mux := dns.NewServeMux()
+	mux.HandleFunc(srv.Domain, api.ServiceLookup)
+	if len(srv.Recursors) > 0 {
+		mux.HandleFunc(".", api.Recurse)
+	}
+
+	errors := make(chan error, 4)
```

Comment:
I guess this has something to do with 2 servers x 2 statuses, but I think this would make more sense as a channel per server, then selecting over them.